mbedtls: fix gcc-11 build by disabling a warning

trofi opened the pull request:
The build fails on gcc-11. Example reproducer:

    $ nix-build -E 'with import ./. {}; mbedtls.override { stdenv = gcc11Stdenv; }'
    ...
    /build/source/library/ssl_tls.c:3334:5: error:
      'mbedtls_sha512_finish_ret' accessing 64 bytes in a region
        of size 48 [-Werror=stringop-overflow=]
     3334 |     finish( &sha512, padbuf );
          |     ^~~~~~~~~~~~~~~~~~~~~~~~~

Instead of backporting an API change let's disable warning for now.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

trofi committed:
mbedtls: fix gcc-11 build by disabling warnings

The build fails on gcc-11. Example reproducer:

    $ nix-build -E 'with import ./. {}; mbedtls.override { stdenv = gcc11Stdenv; }'
    ...
    /build/source/library/ssl_tls.c:3334:5: error:
      'mbedtls_sha512_finish_ret' accessing 64 bytes in a region
        of size 48 [-Werror=stringop-overflow=]
     3334 |     finish( &sha512, padbuf );
          |     ^~~~~~~~~~~~~~~~~~~~~~~~~

Instead of backporting an API change let's disable warnings.

r-burns commented:
I would be in favor of patching out -Werror. Breaking the build on what is essentially a linter error is a waste of time.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 138202` at e907e6d8 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages skipped due to time constraints:</summary>
  <ul>
    <li>belle-sip</li>
    <li>dolphinEmu</li>
    <li>dolphinEmuMaster</li>
    <li>haxePackages.hxcs</li>
    <li>haxePackages.hxjava</li>
    <li>haxe_3_2</li>
    <li>haxe_3_4</li>
    <li>haxe_4_0</li>
    <li>haxe_4_1</li>
    <li>liblinphone</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>21 packages built successfully:</summary>
  <ul>
    <li>bctoolbox</li>
    <li>belcard</li>
    <li>belr</li>
    <li>bzrtp</li>
    <li>dislocker</li>
    <li>gauche</li>
    <li>gaucheBootstrap</li>
    <li>haxe (haxe_4_2)</li>
    <li>hxcpp (haxePackages.hxcpp)</li>
    <li>haxePackages.hxnodejs_4</li>
    <li>hiawatha</li>
    <li>mbedtls</li>
    <li>mediastreamer</li>
    <li>mediastreamer-openh264</li>
    <li>msilbc</li>
    <li>neko</li>
    <li>ortp</li>
    <li>privoxy</li>
    <li>shadowsocks-libev</li>
    <li>trx</li>
    <li>yojimbo</li>
  </ul>
</details>

trofi commented:
> I would be in favor of patching out -Werror. Breaking the build on what is essentially a linter error is a waste of time.

Good point! Switched to `-DMBEDTLS_FATAL_WARNINGS=off` `CMakeLists.txt` knob and dropped existing overrides.

r-burns reviewed (approved):
Neat, verified locally that the build works with gcc11. Nice to get a head start on the warning sprint for the next gcc bump :)

Feel free to target master.

trofi commented:
Good idea! Rebased against `master` and fixed freshly introduced comment typo.

risicle commented:
I wouldn't be so confident about disabling `-Werror`. This can throw up notices of subtle e.g. security problems, and the developer turned it on for a reason.

r-burns commented:
Yes, that's a good reason to promote security-related warnings such as -Werror=format-security for hardening, but blanket -Werror brings too many false positives.
